Add ASN numbers as scan targets

liquidsec opened the pull request:
## Summary

- **ASN as targets**: You can now pass ASN numbers (e.g. `AS13335`) as scan targets. They are automatically expanded into their constituent IP ranges before the scan begins.
- **Migrated to `asndb` library**: ASN lookups now use the [`asndb`](https://github.com/blacklanternsecurity/asndb) library instead of hand-rolled API calls. This handles ASN→CIDR expansion, IP→ASN reverse lookups, and caching.
- **Migrated to `radixtarget` 4.x**: `BBOTTarget` now uses composition (owns a `RadixTarget`) instead of inheritance. This aligns with the radixtarget 4.x string-only API.
- **Replaced SHA-1 hashing with seahash/FNV-1a**: Target and blacklist hashing now uses faster 64-bit hash functions instead of SHA-1.
- **`BBOTTarget.generate_children()`**: New async method on the target object that expands compound seed types (like ASNs) into child events before the scan begins.
- **`shodan_enterprise` module cleanup**: Removed redundant ASN emission (handled by `asn` report module via asndb). Fixed scope filtering to use framework's `in_scope_only` attribute properly.
- **Engine subprocess leak fix**: Fixed a bug where DNS/web engine subprocesses were not properly shut down, causing resource accumulation across tests.
- **`install_core_deps` performance**: Core dependency installation now caches properly, reducing redundant 15s Ansible invocations to ~0s on subsequent runs.
- **Graceful handling for missing module deps**: Modules with uninstallable dependencies no longer crash the scan; they are skipped with a warning. Non-TTY sudo prompts are also handled gracefully.
- **Test fixes**: Fixed several test reliability issues including `test_python_api` preset checks and scope accuracy tests. Added scope filtering tests for `shodan_enterprise`. Removed mocks in favor of real `asndb` lookups.

TheTechromancer commented:
This PR is currently waiting on two things:
- The public release of `asndb.api.bbot.io` (which in turn is waiting on a pentest before it goes live)
- Integration with the new Python client library:
    - https://github.com/blacklanternsecurity/asndb

liquidsec committed:
add ASN numbers as scan targets with async expansion via new ASN helper

liquidsec committed:
Merge branch '3.0' into asn-as-targets

liquidsec committed:
Merge branch '3.0' into asn-as-targets

liquidsec committed:
Merge branch '3.0' into asn-as-targets

liquidsec committed:
fix ASN target len() overflow and JSON serialization errors

liquidsec committed:
store ASN event data as dict for proper JSON round-trip

liquidsec committed:
fix ASN event display in logs and discovery context

liquidsec committed:
remove stale whitelist references

liquidsec committed:
adding the goodest boy

liquidsec committed:
Merge branch '3.0' into asn-as-targets

liquidsec committed:
Merge branch '3.0' into asn-as-targets

liquidsec committed:
lint

TheTechromancer commented on bbot/core/event/base.py:
```diff
@@ -605,7 +605,7 @@ def parent(self, parent):
                 event_has_url = getattr(self, "parsed_url", None) is not None
                 for t in parent.tags:
-                    if t in ("affiliate",):
+                    if t in ("affiliate"):
```
this comma is required for it to be a tuple

TheTechromancer commented on bbot/core/helpers/asn.py:
Most of this will be replaced by the asndb pypi library, which currently needs API key + tests passing in order to publish.

TheTechromancer commented on bbot/core/helpers/misc.py:
```diff
@@ -2595,30 +2597,101 @@ def parse_port_string(port_string):
-async def as_completed(coros):
+async def as_completed(
+    coroutines: Iterable[Awaitable],
+    max_concurrent: Optional[int] = None,
```
Default limit should be something reasonable like 20-30 etc.

TheTechromancer replied:
this still applies correct?

TheTechromancer commented on bbot/core/helpers/simhash.py:
```diff
@@ -0,0 +1,113 @@
+        width = 3
+        text = text.lower()
+        # Remove non-word characters
+        text = re.sub(r"[^\w]+", "", text)
```
this should be precompiled

TheTechromancer commented on bbot/core/modules.py:
```diff
@@ -476,9 +476,10 @@ def load_modules(self, module_names):
                     f"Error loading module {module_name}: {e}. You may have leftover artifacts from an older version of BBOT. Try deleting/renaming your '~/.bbot' directory."
-                ) from e
+                )
+                module = None
```
The error was intentional to stop the scan if a module failed to load

TheTechromancer commented on bbot/scanner/preset/preset.py:
```diff
@@ -290,20 +290,18 @@ def bbot_home(self):
     def blacklist(self):
         if self._target is None:
-            raise ValueError("Cannot access blacklist before preset is baked (use ._blacklist instead)")
+            return None
```
these errors are designed to catch situations where these attributes are accessed prematurely (i.e. before bake). if they're returning None, we shouldn't need to access them

let's put it back and see what breaks

TheTechromancer commented on bbot/scanner/scanner.py:
```diff
@@ -1165,8 +1213,12 @@ def json(self):
+            j["preset"] = self.preset.to_dict(redact_secrets=True)
+        else:
+            j["target"] = {}
+            j["preset"] = {}
```
can we try reverting this to see why it had to be introduced? a scan should never exist without a preset (it should be impossible)

TheTechromancer commented on bbot/scanner/target.py:
after migrating to asndb pypi library, passing in helpers shouldn't be necessary

TheTechromancer reviewed (commented):
asdf

TheTechromancer commented on bbot/core/helpers/command.py:
```diff
@@ -195,7 +195,7 @@ async def _spawn_proc(self, *command, **kwargs):
         kwargs["stdin"] = asyncio.subprocess.PIPE
 
-    log.hugeverbose(f"run: {' '.join(command)}")
+    log.debug(f"run: {' '.join(command)}")
```
we must show commands if -v is specified

liquidsec replied:
reverted

TheTechromancer commented on bbot/core/helpers/command.py:
```diff
@@ -195,7 +195,7 @@ async def _spawn_proc(self, *command, **kwargs):
         kwargs["stdin"] = asyncio.subprocess.PIPE
 
-    log.hugeverbose(f"run: {' '.join(command)}")
+    log.debug(f"run: {' '.join(command)}")
```
we must show commands when -v. if necessary, we can add a exclusion flag

TheTechromancer reviewed (commented):
asdf

liquidsec committed:
Migrate ASN helper to asndb library and address PR review comments

liquidsec committed:
Migrate to radixtarget 4.x with composition pattern

- BaseTarget no longer subclasses RadixTarget; uses composition instead
- Rename strict_dns_scope -> strict_scope everywhere
- Update host_size_key import path for radixtarget 4.x
- Handle radixtarget 4.x API changes (strings-only, no _add, hash is int)
- Skip acl_mode when strict_scope is True (mutually exclusive in 4.x)
- Update test assertions for new hash values and string-based hosts

Work in progress - more test fixes needed.

liquidsec committed:
Fix asndb API calls and radixtarget 4.x string-only compatibility

liquidsec committed:
lint

liquidsec committed:
format

liquidsec committed:
Replace SHA1 hashing with seahash and FNV-1a 64-bit

liquidsec committed:
Fix test_scan hosts assertion for radixtarget 4.x string-only API

liquidsec committed:
Fix portscan passing IPv4Address to radixtarget 4.x search()

liquidsec committed:
Merge branch '3.0' into asn-as-targets

# Conflicts:
#	bbot/scanner/scanner.py
#	bbot/test/test_step_1/test_python_api.py

liquidsec committed:
Merge branch 'asn-as-targets' into migrate-to-asndb-library

liquidsec committed:
ruff format scanner.py

liquidsec committed:
Fix shodan_enterprise: remove redundant ASN emission, use framework scope filtering

- Remove ASN event emission (already handled by asn report module via asndb)
- Replace scope_distance_modifier+filter_event with proper in_scope_only class attr
- Dynamically set scope_distance_modifier=1 when in_scope_only option is disabled
- Add tests for in_scope_only=True and in_scope_only=False behavior

liquidsec committed:
Remove asndb mock from test_target_basic, use real API

liquidsec committed:
Merge pull request #2957 from blacklanternsecurity/migrate-to-asndb-library

Migrate to asndb library

liquidsec committed:
Merge branch '3.0' into asn-as-targets

TheTechromancer commented on bbot/scanner/target.py:
```diff
@@ -341,7 +391,38 @@ def in_target(self, host):
+            expanded_seed_hosts = set(self.seeds.hosts)
+            for host in expanded_seed_hosts:
+                if host not in self.target:
+                    self.target.add(host)
```
Isn't a target by nature custom? I.e. I don't think it's possible to specify ASNs without a custom target? It seems like this should always happen. What do you think

liquidsec replied:
I think the logic is correct (user can assign a more specific/explicit target than the seeds) but the naming made it confusing. Updated variable names/comments to alleviate.

TheTechromancer commented on bbot/test/test_step_2/module_tests/test_module_asn.py:
```diff
@@ -1,239 +1,55 @@
+        asn_helper = module_test.scan.helpers.asn
+        asn_helper._client = AsyncMock()
+        asn_helper._client.lookup_ip = AsyncMock(return_value=self.asndb_response)
+        asn_helper._client.lookup_asn = AsyncMock(return_value=self.asndb_response)
```
Instead of using asyncmock, we should use the real thing. we can hardcode 1.1.1.1 and 8.8.8.8, etc.

liquidsec replied:
I agree, but the problem is the google ASN is ENORMOUS and the test takes too long and times out

liquidsec replied:
and the purpose of the test is to test the module - not the asn behavior itself

liquidsec replied:
nevermind there was a big in the tests which i fixed. removed the mock

TheTechromancer commented on bbot/test/conftest.py:
```diff
@@ -343,6 +343,25 @@ def pytest_sessionfinish(session, exitstatus):
     # Wipe out BBOT home dir
     shutil.rmtree("/tmp/.bbot_test", ignore_errors=True)
 
+    # Ensure stdout/stderr are blocking before pytest writes summaries
+    try:
+        import sys
+        import fcntl
+        import os
+        import io
+
+        fds = []
+        for stream in (sys.stdout, sys.stderr):
+            try:
+                fds.append(stream.fileno())
+            except io.UnsupportedOperation:
+                pass
+        for fd in fds:
+            flags = fcntl.fcntl(fd, fcntl.F_GETFL)
+            fcntl.fcntl(fd, fcntl.F_SETFL, flags & ~os.O_NONBLOCK)
+    except Exception:
+        pass
+
```
was this to address a specific bug?

liquidsec replied:
fixes a bug where pytest crashes during test summary output with BlockingIOError: [Errno 11] write could not complete without blocking.

From claude:

```
 BBOT's engine subprocesses set stdout/stderr to non-blocking mode (via asyncio). If a subprocess doesn't clean up properly on shutdown, the file descriptors stay non-blocking. When pytest then tries to write its   
  summary (which can be large), the write fails because the fd can't accept all the data at once in non-blocking mode.
                                                                                                                                                                                                                        
  It was showing up as flaky test failures where all tests pass but pytest itself crashes at the end trying to print results.                                                                                           
   
```

liquidsec committed:
Clarify seed expansion scope guard naming and comments

liquidsec committed:
Use real asndb API in ASN module test instead of mocks

liquidsec committed:
Default as_completed max_concurrent to 20

liquidsec committed:
Remove mock from unknown ASN test, use private IP instead

liquidsec committed:
ruff format conftest.py
